### PR TITLE
Feature/update frozen

### PR DESCRIPTION
**Pull request by MBkkt**

### CODE REVIEW

File: external/frozen/include/frozen/bits/pmh.h [hunk 1]
```diff
@@ -94,18 +95,23 @@ pmh_buckets<M> constexpr make_pmh_buckets(const carray<Item, N> & items,
...
       b.clear();
     }
     result.seed = prg();
+    rejected = false;
     for (std::size_t i = 0; i < N; ++i) {
       auto & bucket = result.buckets[hash(key(items[i]), static_cast<size_t>(result.seed)) % M];
-      if (bucket.size() >= result_t::bucket_max) { continue; }
+      if (bucket.size() >= result_t::bucket_max) {
+        rejected = true;
+        break;
```
Comment (MBkkt) [inline external/frozen/include/frozen/bits/pmh.h:110]: [Bug fix](https://github.com/serge-sans-paille/frozen/commit/05988d951de3930f140f6ef1101121408794214a#diff-13defd85533f295189ec25d474bf78b9f17c70336e7ddfc54f6f39d3a47be615)

File: external/frozen/include/frozen/random.h [hunk 2]
```diff
@@ -51,19 +62,15 @@ class linear_congruential_engine {
 	  using uint_least_t = bits::select_uint_least_t<bits::log(a) + bits::log(m) + 4>;
     uint_least_t tmp = static_cast<uint_least_t>(multiplier) * state_ + increment;
 
-    // the static cast below may end up doing a truncation
-    if(modulus != 0)
-      state_ = static_cast<result_type>(tmp % modulus);
-    else
-      state_ = static_cast<result_type>(tmp);
+    state_ = modulo(tmp, std::integral_constant<UIntType, modulus>());
```
Comment (MBkkt) [inline external/frozen/include/frozen/random.h:65]: Compile-time if now

File: external/frozen/include/frozen/string.h [hunk 3]
```diff
@@ -28,27 +28,35 @@
...
 #include <functional>
 
+#include <string_view>
+
 namespace frozen {
 
-class string {
-  char const *data_;
+template <typename _CharT>
+class basic_string {
```
Comment (MBkkt) [inline external/frozen/include/frozen/string.h:36]: More convenient string

File: external/frozen/include/frozen/string.h [hunk 4]
```diff
@@ -71,38 +79,62 @@ class string {
...
+  constexpr std::size_t operator()(basic_string<_CharT> value) const {
     std::size_t d = 5381;
     for (std::size_t i = 0; i < value.size(); ++i)
-      d = d * 33 + value[i];
+      d = d * 33 + static_cast<size_t>(value[i]);
     return d;
   }
-  constexpr std::size_t operator()(string value, std::size_t seed) const {
-    std::size_t d = seed;
+  // https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function
```
Comment (MBkkt) [inline external/frozen/include/frozen/string.h:92]: https://github.com/serge-sans-paille/frozen/commit/30b475274c044f456c37b6d15de3d04c85112698
Better string hash function

**Review by Dronplane — APPROVED** [review]
LGTM But I'd prefer to keep README file and to not modify original defines/usings if it is not absolutely necessary

File: external/frozen/include/frozen/bits/pmh.h [hunk 5]
```diff
@@ -125,7 +131,7 @@ struct seed_or_index {
   using value_type = uint64_t;
 
 private:
-  static constexpr value_type MINUS_ONE = -1;
+  static constexpr value_type MINUS_ONE = std::numeric_limits<value_type>::max();
```
Comment (alexbakharew) [inline external/frozen/include/frozen/bits/pmh.h:134]: Why -1 become std::numeric_limits<value_type>::max(); ?

File: external/frozen/include/frozen/bits/pmh.h [hunk 6]
```diff
@@ -177,7 +183,7 @@ pmh_tables<M, Hash> constexpr make_pmh_tables(const carray<Item, N> &
   carray<seed_or_index, M> G; // Default constructed to "index 0"
 
   // H becomes the second hash table in the resulting pmh function
-  constexpr std::size_t UNUSED = -1;
+  constexpr std::size_t UNUSED = std::numeric_limits<std::size_t>::max();
```
Comment (alexbakharew) [inline external/frozen/include/frozen/bits/pmh.h:186]: same

**Review by alexbakharew — APPROVED** [review]

File: external/frozen/include/frozen/bits/pmh.h (hunk 5 above)
Comment (MBkkt) [reply external/frozen/include/frozen/bits/pmh.h:134]: It's their code. I think value_type is size_t, and they want avoid warning

File: external/frozen/include/frozen/bits/pmh.h (hunk 6 above)
Comment (MBkkt) [reply external/frozen/include/frozen/bits/pmh.h:186]: same here

**Pull request merged**